feat: expose attestation type and trust path after validation

bdewater opened the pull request:
The WebAuthn spec [notes](https://www.w3.org/TR/2018/CR-webauthn-20180807/#sctn-attestation-types) the following:

> Attestation statements conveying attestations of type AttCA use the same data structure as attestation statements conveying attestations of type Basic, so the two attestation types are, in general, distinguishable only with externally provided knowledge regarding the contents of the attestation certificates conveyed in the attestation

[Retrieving](https://fidoalliance.org/mds/) and managing the this external data ([TOC](https://fidoalliance.org/specs/fido-v2.0-id-20180227/fido-metadata-service-v2.0-id-20180227.html) and [statements](https://fidoalliance.org/specs/fido-v2.0-id-20180227/fido-metadata-statement-v2.0-id-20180227.html)) from the FIDO Metadata Service is up to the implementing application (for now? https://github.com/cedarcode/webauthn-ruby/issues/66) if so desired.

Closes https://github.com/cedarcode/webauthn-ruby/issues/80

grzuy commented on lib/webauthn/attestation_statement/packed.rb:
```diff
@@ -86,6 +87,16 @@ def valid_signature?(authenticator_data, client_data_hash)
+        if raw_attestation_certificates&.any?
+          [ATTESTATION_TYPE_BASIC_OR_ATTCA, attestation_certificate_chain]
+        elsif raw_ecdaa_key_id
+          [ATTESTATION_TYPE_ECDAA, raw_ecdaa_key_id]
```
We don't support `ECDAA` for now.
Can we please remove it from here, it will never execute this code.

grzuy commented on lib/webauthn/attestation_statement/base.rb:
```diff
@@ -5,6 +5,13 @@ module AttestationStatement
+      ATTESTATION_TYPE_SELF = "self"
+      ATTESTATION_TYPE_ATTCA = "attca"
+      ATTESTATION_TYPE_ECDAA = "ecdaa"
+      ATTESTATION_TYPE_BASIC_OR_ATTCA = "basic_or_attca"
```
W3C Recommendation seem to use first letter uppercase, e.g. `None` instead of `none` or `AttCA` instead of `attca`.

grzuy replied:
Also, can we move these out of `Base` namespace?

grzuy commented on spec/webauthn/authenticator_attestation_response_spec.rb:
```diff
@@ -39,6 +39,8 @@
 
     expect(response.valid?(original_challenge, original_origin)).to eq(true)
+    expect(response.attestation_type).to eq(WebAuthn::AttestationStatement::Base::ATTESTATION_TYPE_BASIC_OR_ATTCA)
+    expect(response.attestation_trust_path).to be_kind_of(OpenSSL::X509::Certificate)
```
From what I understand W3C recommendations states it should return `x5c` which stands for a 1 element array, not the actual element.
https://www.w3.org/TR/2018/CR-webauthn-20180807/#fido-u2f-attestation

grzuy commented on spec/webauthn/authenticator_attestation_response_spec.rb:
```diff
@@ -39,6 +39,8 @@
     )
 
     expect(response.valid?(original_challenge, original_origin)).to eq(true)
+    expect(response.attestation_type).to eq(WebAuthn::AttestationStatement::Base::ATTESTATION_TYPE_BASIC_OR_ATTCA)
```
I only see "Basic" as the returned type in the "Verification procedure" in https://www.w3.org/TR/2018/CR-webauthn-20180807/#fido-u2f-attestation. I may be missing something.

grzuy commented on spec/webauthn/authenticator_attestation_response_spec.rb:
```diff
@@ -55,6 +57,8 @@
     )
 
     expect(response.valid?(original_challenge, original_origin)).to eq(true)
+    expect(response.attestation_type).to eq(WebAuthn::AttestationStatement::Base::ATTESTATION_TYPE_SELF)
```
I would try to avoid using internal constants for test expectations.
We want the test to fail if we accidentally change the constant value, right?

bdewater commented on spec/webauthn/attestation_statement/packed_spec.rb:
```diff
@@ -24,11 +24,8 @@
-      )
-    )
+    attestation_type, attestation_trust_path = subject.valid?(authenticator_data, client_data_hash)
+    expect(attestation_type).to eq(WebAuthn::AttestationStatement::Base::ATTESTATION_TYPE_BASIC_OR_ATTCA)
```
This has been [clarified](https://github.com/w3c/webauthn/commit/a05982cbd8ce4ba3b0e4c4e18f628a8de5cbe4da) in the latest editors draft. In the version you're linking, signing procedure step 2 does mention "If Basic or AttCA attestation is in use..."

grzuy replied:
> This has been clarified in the latest editors draft

Gotcha :+1:

grzuy commented:
@bdewater Want me to help out with the changes?

grzuy commented:
...or help you in any other way?

bdewater commented:
Hi @grzuy I've addressed your comments, please let me know if you'd like to see other changes. Just have been busy the last few weeks, but I intend to get this over the finish line 😄

grzuy commented on lib/webauthn/attestation_statement/none.rb:
```diff
@@ -6,7 +6,7 @@ module WebAuthn
     class None < Base
       def valid?(*_args)
-        true
+        [WebAuthn::AttestationStatement::Types::NONE, nil]
```
I think with this new `Types` namespace I find reading this a bit confusing.

I am worried that reading "attestation statement type none" would confuse users into thinking this is talking about "attestation statement formats", which is a different thing.

I more inclined to do `WebAuthn::AttestationFormat::ATTESTATION_TYPE_NONE` or even `WebAuthn::ATTESTATION_TYPE_NONE`.

bdewater replied:
I like `WebAuthn::AttestationFormat::ATTESTATION_TYPE_NONE`, I think it captures well how [Attestation Statement Formats part of the spec](https://www.w3.org/TR/webauthn/#attestation-formats) describes the relation between the two. Thanks, I've updated the PR!

grzuy commented:
Awesome, thank you very much!

Added additional commit to fix tests.
Merging now.
